src/libraries/Common/src/System/Net/Security/CertificateValidation.OSX.cs

```diff
@@ -0,0 +1,75 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.Win32.SafeHandles;
+using System.Diagnostics;
+using System.Globalization;
+using System.Runtime.InteropServices;
+using System.Security.Cryptography.X509Certificates;
+
+namespace System.Net.Security
+{
+    internal static class CertificateValidation
+    {
+        private static readonly IdnMapping s_idnMapping = new IdnMapping();
+
+        // WARNING: This function will do the verification using OpenSSL. If the intention is to use OS function, caller should use CertificatePal interface.
+        internal static SslPolicyErrors BuildChainAndVerifyProperties(X509Chain chain, X509Certificate2 remoteCertificate, bool checkCertName, bool isServer, string? hostName, IntPtr certificateBuffer, int bufferLength = 0)
+        {
+            SslPolicyErrors errors = chain.Build(remoteCertificate) ?
+                SslPolicyErrors.None :
+                SslPolicyErrors.RemoteCertificateChainErrors;
+
+            if (!checkCertName)
+            {
+                return errors;
+            }
+
+            if (string.IsNullOrEmpty(hostName))
+            {
+                return errors | SslPolicyErrors.RemoteCertificateNameMismatch;
+            }
+
+            SafeX509Handle certHandle;
+            if (certificateBuffer != IntPtr.Zero && bufferLength > 0)
+            {
+                certHandle = Interop.Crypto.DecodeX509(certificateBuffer, bufferLength);
+            }
+            else
+            {
+                // We dont't have DER encoded buffer.
+                byte[] der = remoteCertificate.Export(X509ContentType.Cert);
```

Comment:
The Handle is pointer to native memory that represents the certificate it self. So on macOS it would be `SecCertificateRef` or `SecKeychainItemRef` We cannot pass that directly to OpenSSL. It works on Linux because the handle is already `X509*` e.g. object created by OpenSSL itself.

Reply:
On mac, the cert handle should be an X509*. All the openssl based backends that should be the case, unless youre explicitly passing QUIC_CREDENTIAL_FLAGS_USE_PORTABLE_CERTIFICATES